projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,267 @@
+    std::string csvpath = folderName + "/coordinates.csv";
+    coordinates.open (csvpath, std::ofstream::trunc);
+    coordinates << "speed" << "," << "steering angle" << "," << "gps quality" << "," << "lat" << "," << "lon" << "," << "altitude"  << "," << "roll" << "," << "pitch" << "," << "yaw"  << ",";
+    coordinates << "image name" << "," << "timestamp" <<"\n";
+    coordinates.close();
+
+
+    BoBRobotics::BackgroundExceptionCatcher catcher;
+    catcher.trapSignals();
+
+    int i = 0;
+    BoBRobotics::Stopwatch sw, sw_timestamp;
+    std::array<degree_t, 3> angles;
+    // for x timestep
+    sw_timestamp.start();
+    for (;;) {
+        // poll from camera thread
+
+
+
+        cv::Mat originalImage;
+        mex.lock();
+        tmpImage.copyTo(originalImage);
+        mex.unlock();
+
+        std::ostringstream fileString, folderString;
+        fileString << folderName << "/image" << i << ".jpg";
+        std::string fileName = fileString.str();
+        std::thread imwriteThread{ &saveThread, std::move(originalImage), std::move(fileName) };
+        imwriteThread.detach();
+
+        try {
+            catcher.check();
+            sw.start();
+            // get imu data
+            angles = imu.getEulerAngles();
+            float yaw = angles[0].value();
+            float pitch = angles[1].value();
+            float roll = angles[2].value();
+
+            // get gps data
+            BoBRobotics::GPS::GPSData data = gps.getGPSData();
+            BoBRobotics::MapCoordinate::GPSCoordinate coord = data.coordinate;
+            BoBRobotics::GPS::TimeStamp time = data.time;
+            int gpsQual = (int) data.gpsQuality; // gps quality
+
+            // output results
+    	   // if (i%5 == 0) { // only output once ever sec
+            std::cout << std::setprecision(10) << "GPS quality: " << gpsQual << " latitude: " << coord.lat.value()<< " longitude: " <<  coord.lon.value()  << " num sats: " << data.numberOfSatellites << " time: " << static_cast<units::time::millisecond_t>(sw_timestamp.elapsed()).value()/1000 << std::endl;
+
+
+	        folderString << folderName << "/coordinates.csv";
+
+
+            bot.updateState();
+            auto bot_speed = bot.getSpeed();
+            auto turn_ang = bot.getTurningAngle();
+
+            //  headers:   speed|steering_angle|gpsQuality|latitude|longitude|altitude|roll|pitch|yaw|imagename|timestamp
+            coordinates.open (folderString.str(), std::ofstream::app); // open coordinates.csv file and append
+            coordinates << std::setprecision(10) << std::fixed << bot_speed << "," <<  turn_ang.to<double>() << "," << gpsQual << "," << coord.lat.value() << "," << coord.lon.value() << "," << data.altitude.value()  << "," << roll << "," << pitch << "," << yaw  << ",";
+            coordinates << "image" << i << ".jpg" << "," << static_cast<units::time::millisecond_t>(sw_timestamp.elapsed()).value() <<"\n";
+            coordinates.close();
+
+
+
+
+
+
+
```

Comment:
*Definitely* not important for today, but just a reminder that you can get VS code to clean up your whitespace by right-clicking and choosing Format document.